### DISCUSSION TITLE
Recommandation for using RQ with fork unsafe libraries

### DISCUSSION THREAD

**Discussion by stripedpumpkin** [rq/rq#1766] [Q&A]
Hello. I was wondering what the recommandation would be to use RQ with fork unsafe libraries such as PyMongo (https://pymongo.readthedocs.io/en/stable/faq.html#is-pymongo-fork-safe).

The use case would be the following: accessing a mongo database within the execution of an RQ job.

In the current case:
- either the client (== pool) is initialize pre-fork => that leads to errors,
- it is initialized post-fork, but then we are initializing a fresh connection with each job, which is not ideal as this limits scalability and adds a lot of overhead to each job execution.

(If it matters this setup would happen in the context of Django app.)

Would you recommend using another parallelization mechanism ? If so what would be the go to reference and how to set this up ?

Thanks for reading

**Comment by ccrvlh** [comment]
@stripedpumpkin I don't have an easy solution, I'm not particular familiar with how `pymongo` handles connection/pooling etc. Maybe more details on why grabbing connections from the `pool` wont work could be helpful to try to implement something fork-safe?

You could customize the `Worker` class though to try to make it work with `pymongo`. The forking happens here:

https://github.com/rq/rq/blob/c5a7a6b670582a20e353d5c5322304bf9ff1057c/rq/worker.py#L795-L807

I think you could get way with doing something similar, but instead of forking, creating a process manually. You might need to pass some things that are on the parent process manually though: I found [this article](https://britishgeologicalsurvey.github.io/science/python-forking-vs-spawn/) interesting when reading about this specific topic. 

You would probably also need to handle this method:

https://github.com/rq/rq/blob/c5a7a6b670582a20e353d5c5322304bf9ff1057c/rq/worker.py#L919-L935

To be honest, I'm not sure if that would work at all, but might be worth the shot.
I'm moving this to the discussions, seems more appropriate as it's not actually an issue when considering the way `RQ` works. Feel free to follow up on this, would be great to hear how you handle this case, could be helpful to others as well.

**Reply by ccrvlh** [reply]
@stripedpumpkin I completely missed the `SimpleWorker` which does not `fork()` and runs jobs in the same process, perhaps you can give that a try as well.
